sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/TimeWindow.scala

```diff
@@ -0,0 +1,137 @@
+        return TypeCheckFailure(s"The window duration ($windowDuration) must be greater than 0.")
+      }
+      if (slideDuration <= 0) {
+        return TypeCheckFailure(s"The slide duration ($slideDuration) must be greater than 0.")
+      }
+      if (startTime < 0) {
+        return TypeCheckFailure(s"The start time ($startTime) must be greater than or equal to 0.")
+      }
+      if (slideDuration > windowDuration) {
+        return TypeCheckFailure(s"The slide duration ($slideDuration) must be less than or equal to the " +
+            s"windowDuration ($windowDuration).")
+      }
+      if (startTime >= slideDuration) {
+        return TypeCheckFailure(s"The start time ($startTime) must be less than the " +
+            s"slideDuration ($slideDuration).")
+      }
+      return dataTypeCheck
+    } else {
+      return dataTypeCheck
+    }
+  }
+  /**
+   * Validate the inputs for the window duration, slide duration, and start time.
+   *
+   * @return Some string with a useful error message for the invalid input.
+   */
+  def validate(): Option[String] = {
+    if (windowDuration <= 0) {
+      return Some(s"The window duration ($windowDuration) must be greater than 0.")
+    }
+    if (slideDuration <= 0) {
+      return Some(s"The slide duration ($slideDuration) must be greater than 0.")
+    }
+    if (startTime < 0) {
+      return Some(s"The start time ($startTime) must be greater than or equal to 0.")
+    }
+    if (slideDuration > windowDuration) {
+      return Some(s"The slide duration ($slideDuration) must be less than or equal to the " +
+        s"windowDuration ($windowDuration).")
+    }
+    if (startTime >= slideDuration) {
+      return Some(s"The start time ($startTime) must be less than the " +
+        s"slideDuration ($slideDuration).")
+    }
+    None
+  }
+}
+
+object TimeWindow {
+  /**
+   * Parses the interval string for a valid time duration. CalendarInterval expects interval
+   * strings to start with the string `interval`. For usability, we prepend `interval` to the string
+   * if the user ommitted it.
+   *
+   * @param interval The interval string
+   * @return The interval duration in seconds. SparkSQL casts TimestampType to Long in seconds,
+   *         therefore we use seconds here as well.
+   */
+  private def getIntervalInSeconds(interval: String): Long = {
+    if (StringUtils.isBlank(interval)) {
+      throw new IllegalArgumentException(
+        "The window duration, slide duration and start time cannot be null or blank.")
+    }
+    val intervalString = if (interval.startsWith("interval")) {
+      interval
+    } else {
+      "interval " + interval
+    }
+    val cal = CalendarInterval.fromString(intervalString)
+    if (cal == null) {
+      throw new IllegalArgumentException(
+        s"The provided interval ($interval) did not correspond to a valid interval string.")
+    }
+    (cal.months * 4 * CalendarInterval.MICROS_PER_WEEK + cal.microseconds) / 1000000
```

Comment:
We can't operate on TimestampType as if they are Longs. They get cast to
LongType AFAIK, which has second precision. I'm not sure if supporting ms
is possible.
On Mar 29, 2016 11:14 PM, "Shixiong Zhu" <notifications@github.com> wrote:

> In
> sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/TimeWindow.scala
> <https://github.com/apache/spark/pull/12008#discussion_r57840903>:
>
> > +  private def getIntervalInSeconds(interval: String): Long = {
> > +    if (StringUtils.isBlank(interval)) {
> > +      throw new IllegalArgumentException(
> > +        "The window duration, slide duration and start time cannot be null or blank.")
> > +    }
> > +    val intervalString = if (interval.startsWith("interval")) {
> > +      interval
> > +    } else {
> > +      "interval " + interval
> > +    }
> > +    val cal = CalendarInterval.fromString(intervalString)
> > +    if (cal == null) {
> > +      throw new IllegalArgumentException(
> > +        s"The provided interval ($interval) did not correspond to a valid interval string.")
> > +    }
> > +    (cal.months * 4 * CalendarInterval.MICROS_PER_WEEK + cal.microseconds) / 1000000
>
> By the way, does it mean that the user cannot use window("timestamp", "500
> milliseconds")?
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly or view it on GitHub
> <https://github.com/apache/spark/pull/12008/files/b4e2fc23585413b1bf50e2487437dd38b9cd748f#r57840903>
>